README.md
coffee-processing
=================

Helps writing Processing.js sketches in Coffeescript.

Installation
------------

Add this line to your application's Gemfile:

    gem 'coffee-processing'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install coffee-processing

Usage
-----

A sample Processing.js sketch written in Coffeescript with the help of coffee-processing.
```coffee
setup = ->
  size $(window).width(), $(window).height()
  frameRate 30
  background 255

draw = ->
  for i in [0..10]
    s = random(100)
    stroke random(255), random(255), random(255)
    ellipse random(width()), random(height()), s, s
```

Ruby code for compiling it into Javascript.
```ruby
require 'coffee-processing'

File.open('compiled.js', 'w') do |f|
  f << CoffeeProcessing.compile('this.sketch', code)
end
```

And the HTML page.
```html
<script src='compiled.js' type='text/javascript'></script>
<canvas id='sketch'></canvas>
<script type='text/javascript'>
  var processing = new Processing(document.getElementById("sketch"), this.sketch)
</script>
```

Preloading (Experimental)
-------------------------

Asynchronous nature of Javascript requires preloading of images and fonts.
Processing.js allows you to preload assets with [@pjs directives](http://processingjs.org/reference/pjs%20directive/).
However they are only applied during the compilation of Processing syntax,
and there's currenty no direct way to preload in Javascript-only Processing.js code. 

If you define `preload` object in your coffee-processing code as the following example,
coffee-processing will delay the execution until those assets are ready.
(Although the way it does so seems quite hacky.)

```coffee
preload =
  images: ['/images/image1.png', '/images/image2.png']
  fonts:  ['/fonts/font.ttf']

setup = ->
  font = createFont '/fonts/font.ttf', 0 
  img1 = loadImage '/images/image1.png'
  img2 = loadImage '/images/image2.png'
```

coffee-processing script
------------------------

```
usage: coffee-processing [--template] <js object name> <sketch file>

        --template                   Create a template page for the sketch
    -h, --help                       Show this message
```


Caveats
-------

Non-constant, instance variables of Processing object,
such as `width`, `frameCount` and `__mousePressed` (among others)
should be accessed through their corresponding shortcut functions as follows.

```coffee
# __mousePressed, frameCount, width and height are functions, not values.

setup = ->
  size 100, 100, P3D

draw = ->
  if __mousePressed()
    point frameCount() % width(), frameCount() % height()
```

Or equivalently, you can access them as the properties of `processing` object.
This is slightly more efficient.

```coffee
# Alias for processing instance
p5 = processing

setup = ->
  size 100, 100, P3D

draw = ->
  if p5.__mousePressed
    point p5.frameCount % p5.width, p5.frameCount % p5.height
```

Examples
--------

Check out [examples](https://github.com/junegunn/coffee-processing/tree/master/examples) directory.

Also check out [a Sinatra demo app](https://github.com/junegunn/coffee-processing-live).

Contributing
------------

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
0.0.6
=====
- Minor bug fix

0.0.5
=====
- Experimental image/font preloading.

0.0.4
=====
- Correctly adjusts line number in parse error message

0.0.3
=====
- Constants (e.g. `P3D`, `RGB`) are now directly accessible without `processing` variable or shortcut methods.
- The same holds true for `ArrayList`, `HashMap`, `PVector`, `ObjectIterator` and `PConstants`

0.0.2
=====
- Prevents accidental assignment of `processing` variable which breaks the sketch.

```coffee
# This used to break the whole sketch
processing = null

setup = -> # ...
```



test/bin_helper.rb
#!/usr/bin/env ruby
# encoding: utf-8
# gunn@daumcorp.com

# DISCLAIMER: 
#   Not a real test!
#   Just a helper script for running scripts with local source

$LOAD_PATH.unshift File.join(File.dirname(__FILE__), '../lib')
load File.join(File.dirname(__FILE__), '../bin/', File.basename(ARGV.shift))


test/test_coffee_processing.rb
#!/usr/bin/env ruby

$LOAD_PATH.unshift File.join(File.dirname(__FILE__), '../lib')
require 'test-unit'
require 'coffee-processing'

class TestCoffeeProcessing < Test::Unit::TestCase
  def test_indentation
    without_indentation = "
setup = ->
  null

draw = ->
  null"

    with_space_indentation = "
      setup = ->
        null

      draw = ->
        null
    "

    with_tab_indentation = "
			setup = ->
				null

			draw = ->
				null
		"

    assert_equal 1, 
      [without_indentation, with_space_indentation, with_tab_indentation].map { |code|
        CoffeeProcessing.compile 'this.sketch', code
      }.uniq.length
  end

  def test_error
    begin
      CoffeeProcessing.compile 'this.sketch', %w[1 2 3 ? 5].join($/)
      assert false, 'Exception should have been raised'
    rescue ExecJS::ProgramError => e
      assert e.to_s =~ /line 4:/, 'Unadjusted line number'
    end
  end

  def test_preload_font
    require 'fileutils'
    code = "
preload =
  fonts: ['data/Bountiful.ttf']
  images: ['data/header.png']

font = null

setup = ->
  size 300, 300
  frameRate 1
  background 255
  image loadImage('data/header.png'), 0, 0
  fill 0
  font = createFont('data/Bountiful.ttf', 0)
  textFont font, 30
  text 'Hello world', 50, 50

mousePressed = ->
  background random 255
  fill random 255
  text 'Hello world', random(50), random(50)

draw = ->
  console.log 'draw'
"
    output = File.join(File.dirname(__FILE__), 'output')
    output_f = File.join(output, 'data')
    FileUtils.mkdir_p output_f
    FileUtils.cp File.join(File.dirname(__FILE__), 'data/Bountiful.ttf'), output_f
    CoffeeProcessing.generate_template_page 'this.sketch', code, output
  end
end


lib/coffee-processing.rb
require "coffee-processing/version"
require 'erubis'
require 'coffee-script'

module CoffeeProcessing
  def self.compile javascript_object, code
    first_line = code.lines.map(&:chomp).find { |line| !line.empty? }

    if first_line && indentation = first_line.scan(/^\s+/).first
      code = code.lines.map { |line| line.sub(/^#{indentation}/, '')}.join
    end

    code = code.lines.map { |line| "#{line}  " }.join
    code = code.strip

    begin
      CoffeeScript.compile(
        Erubis::Eruby.new(
          File.read File.join(
            File.dirname(__FILE__),
            'coffee-processing',
            'boilerplate.coffee.erb'
          )
        ).result(binding)
      )
    rescue ExecJS::ProgramError => e
      message = e.message
      if match = message.match(/line ([1-9][0-9]*):/)
        range = Range.new(*match.offset(1), true)
        message[ range ] = (match[1].to_i - 15).to_s
      end
      raise e, message, e.backtrace
    end
  end

  def self.generate_template_page javascript_object, code, output_dir
    require 'fileutils'
    FileUtils.mkdir_p output_dir
    File.open(File.join(output_dir, 'sketch.js'), 'w') do |f|
      f << compile(javascript_object, code)
    end
    File.open(File.join(output_dir, 'index.html'), 'w') do |f|
      f <<
        Erubis::Eruby.new(
          File.read File.join(
            File.dirname(__FILE__),
            'coffee-processing',
            'template.html.erb'
          )
        ).result(binding)
    end
  end
end


examples/lines/sketch.js
(function() {

  this.sketch = function(processing) {
    var callback, count, draw, exag, mousePressed, offset, setup, status, steps, __sym, _i, _len, _ref, _results;
    for (__sym in processing) {
      if (eval("typeof " + __sym + " === 'undefined'")) {
        if (typeof processing[__sym] === 'function') {
          eval("var " + __sym + " = function() { return processing." + __sym + ".apply(processing, arguments) }");
        } else {
          eval("var " + __sym + " = function() { return processing." + __sym + " }");
        }
      }
    }
    null;

    steps = 5;
    count = 200;
    offset = 0;
    status = 0;
    exag = function(v, times) {
      var i, _i;
      for (i = _i = 0; 0 <= times ? _i < times : _i > times; i = 0 <= times ? ++_i : --_i) {
        v = 6 * pow(v, 5) - 15 * pow(v, 4) + 10 * pow(v, 3);
      }
      return v;
    };
    setup = function() {
      size($(window).width(), $(window).height());
      frameRate(30);
      colorMode(RGB(), 1.0);
      background(1);
      return stroke(0, 0.1);
    };
    mousePressed = function() {
      if (status === 0) {
        return status = 1;
      } else {
        offset = frameCount() * steps % width();
        background(1);
        return status = 0;
      }
    };
    draw = function() {
      var h, i, s, w, x, _i, _j;
      w = width();
      h = height();
      if (status !== 2) {
        for (s = _i = 0; 0 <= steps ? _i < steps : _i > steps; s = 0 <= steps ? ++_i : --_i) {
          x = (frameCount() - 1) * steps + s;
          for (i = _j = 0; 0 <= count ? _j < count : _j > count; i = 0 <= count ? ++_j : --_j) {
            point((x - offset) % w, h / 2 + h * (exag(noise(i / count, x / w), 2) - 0.5));
          }
          if (status === 1 && (x - offset) % w === 0) {
            status = 2;
            break;
          }
        }
      }
      return null;
    };
    _ref = ['setup', 'draw', 'mouseClicked', 'mouseDragged', 'mouseMoved', 'mouseOut', 'mouseOver', 'mousePressed', 'mouseReleased', 'keyPressed', 'keyReleased', 'keyTyped'];
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      callback = _ref[_i];
      if (!eval("typeof " + __sym + " === 'undefined'")) {
        _results.push(eval("processing." + callback + " = " + callback));
      }
    }
    return _results;
  };

}).call(this);
